### PR TITLE
Enhance OSSL_HTTP_parse_url() and use this in apps/cmp.c

### PR DESCRIPTION
**Pull request by DDvO**
* `OSSL_HTTP_parse_url()`: add optional port number return parameter and strengthen documentation.
If this parameter is used the function thoroughly checks the port specification.
* apps/cmp.c: Use enhanced `OSSL_HTTP_parse_url()`, removing local `parse_addr()` and `atoint()`
* apps/cmp.c: Allow default HTTP path (aka CMP alias) given with `-server` option

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by DDvO** [comment]
As this constitutes an API extension I ask for a swift review for inclusion before feature freeze.

**Comment by DDvO** [comment]
Ping - can an OTC member please review this soon since it includes a minor API extension.

File: doc/man3/OSSL_HTTP_transfer.pod [hunk 1]
```diff
@@ -190,13 +190,20 @@ Since this function is typically called by applications such as
 L<openssl-s_client(1)> it uses the B<bio_err> and B<prog> parameters (unless
 NULL) to print additional diagnostic information in a user-oriented way.
 
-OSSL_HTTP_parse_url() parses its input string B<url> as a URL and splits it up
+OSSL_HTTP_parse_url() parses its input string B<url> as a URL
+of the form C<[http[s]://]address[:port][/path]> and splits it up
 into host, port and path components and a flag whether it begins with 'https'.
```
Comment (slontis) [inline doc/man3/OSSL_HTTP_transfer.pod:195]: a flag indicating whether

File: doc/man3/OSSL_HTTP_transfer.pod [hunk 2]
```diff
@@ -190,13 +190,20 @@ Since this function is typically called by applications such as
...
-The host component may be a DNS name or an IPv4 or an IPv6 address.
+The host component may be a DNS name or an IPv4 address
+or an IPv6 address enclosed in C<[> and C<]>.
```
Comment (slontis) [inline doc/man3/OSSL_HTTP_transfer.pod:197]: Are all 3 enclosed in [] or just IPv6?  Not sure how to read this.

File: doc/man3/OSSL_HTTP_transfer.pod [hunk 3]
```diff
@@ -190,13 +190,20 @@ Since this function is typically called by applications such as
...
 The port component is optional and defaults to "443" for HTTPS, else "80".
+If the B<pport_num> argument is NULL the port specification
+can be in mnemonic form such as "http" like with L<BIO_set_conn_port(3)>, else
+it must be in numerical form and its integer value is assigned to B<*pport_num>.
 The path component is also optional and defaults to "/".
-As far as the result pointer arguments are not NULL it assigns via
-them copies of the respective string components.
+As far as the result pointer arguments are not NULL the function assigns
```
Comment (slontis) [inline doc/man3/OSSL_HTTP_transfer.pod:203]: This sentence makes not sense.

File: doc/man3/OSSL_HTTP_transfer.pod [hunk 4]
```diff
@@ -190,13 +190,20 @@ Since this function is typically called by applications such as
...
+via them the respective url components.
+For string values these are non-NULL pointers to copies of the strings
+where B<*ppath> is guaranteed to begin with a C</> character.
```
Comment (slontis) [inline doc/man3/OSSL_HTTP_transfer.pod:206]: This 'where ... character' part should not be in the sentence.

File: doc/man3/OSSL_HTTP_transfer.pod [hunk 5]
```diff
@@ -190,13 +190,20 @@ Since this function is typically called by applications such as
...
 The port component is optional and defaults to "443" for HTTPS, else "80".
+If the B<pport_num> argument is NULL the port specification
+can be in mnemonic form such as "http" like with L<BIO_set_conn_port(3)>, else
+it must be in numerical form and its integer value is assigned to B<*pport_num>.
 The path component is also optional and defaults to "/".
-As far as the result pointer arguments are not NULL it assigns via
-them copies of the respective string components.
+As far as the result pointer arguments are not NULL the function assigns
+via them the respective url components.
+For string values these are non-NULL pointers to copies of the strings
```
Comment (slontis) [inline doc/man3/OSSL_HTTP_transfer.pod:205]: 'For strings ...' should be collapsed down and merged into the 'The strings merged this way' part.

File: crypto/http/http_lib.c [hunk 6]
```diff
@@ -21,11 +21,12 @@
...
 {
     char *p, *buf;
-    char *host;
-    const char *port = OSSL_HTTP_PORT;
+    char *host, *host_end;
+    const char *path, *port = OSSL_HTTP_PORT;
+    long portnum = 80;
     size_t https_len = strlen(OSSL_HTTPS_NAME);
 
     if (!ossl_assert(https_len >= strlen(OSSL_HTTP_NAME)))
```
Comment (slontis) [inline crypto/http/http_lib.c:32]: This is a very strange check..

File: crypto/http/http_lib.c [hunk 7]
```diff
@@ -21,11 +21,12 @@
...
 int OSSL_HTTP_parse_url(const char *url, char **phost, char **pport,
-                        char **ppath, int *pssl)
+                        int *pport_num, char **ppath, int *pssl)
 {
     char *p, *buf;
-    char *host;
-    const char *port = OSSL_HTTP_PORT;
+    char *host, *host_end;
+    const char *path, *port = OSSL_HTTP_PORT;
+    long portnum = 80;
```
Comment (slontis) [inline crypto/http/http_lib.c:29]: Why use a long and then return an int?

File: crypto/http/http_lib.c [hunk 8]
```diff
@@ -48,7 +49,7 @@ int OSSL_HTTP_parse_url(const char *url, char **phost, char **pport,
     if ((buf = OPENSSL_strdup(url)) == NULL)
         goto err;
 
-    /* Check for initial colon */
+    /* check for optional http[s]:// prefix */
     p = strchr(buf, ':');
     if (p == NULL || (size_t)(p - buf) > https_len) {
```
Comment (slontis) [inline crypto/http/http_lib.c:54]: It this supposed to be allowed?

anynamebiggerthan5chars://something.com

Is this related to @ ??

File: crypto/http/http_lib.c [hunk 9]
```diff
@@ -59,44 +60,70 @@ int OSSL_HTTP_parse_url(const char *url, char **phost, char **pport,
...
-        p = strchr(host, ']');
-        if (p == NULL)
+        host_end = strchr(host, ']');
+        if (host_end == NULL)
             goto parse_err;
-        *p = '\0';
-        p++;
+        *host_end++ = '\0';
+    } else {
+        host_end = strchr(host, ':');
```
Comment (slontis) [inline crypto/http/http_lib.c:84]: Maybe some comments here
/* search for start of optional port */
...
/* search for start of optional path */
...

/* If there is no port or path so the remaining string is just the hostname */

File: crypto/http/http_lib.c [hunk 10]
```diff
@@ -59,44 +60,70 @@ int OSSL_HTTP_parse_url(const char *url, char **phost, char **pport,
...
+                HTTPerr(0, HTTP_R_INVALID_PORT_NUMBER);
+                goto err;
+            }
+        }
     }
+    *host_end = '\0';
+    *p = '\0'; /* terminate port string */
+
+    /* check for optional path at end of url starting with '/' */
+    path = url + (p - buf);
```
Comment (slontis) [inline crypto/http/http_lib.c:113]: Not sure why you do this..
Could you just have done this?

```
if (*p == '\') {
    *p = '\0';
    path = p + 1;
}
```

File: apps/cmp.c [hunk 11]
```diff
@@ -1883,33 +1839,36 @@ static int handle_opt_geninfo(OSSL_CMP_CTX *ctx)
...
         goto oom;
     if (opt_no_proxy != NULL && !OSSL_CMP_CTX_set1_no_proxy(ctx, opt_no_proxy))
         goto oom;
-    (void)BIO_snprintf(server_buf, sizeof(server_buf), "http%s://%s%s%s/%s",
-                       opt_tls_used ? "s" : "", opt_server,
-                       server_port == 0 ? "" : ":", server_port_s,
-                       opt_path == NULL ? "" :
-                       opt_path[0] == '/' ? opt_path + 1 : opt_path);
+    (void)BIO_snprintf(server_buf, sizeof(server_buf), "http%s://%s:%s/%s",
+                       opt_tls_used ? "s" : "", server, port,
```
Comment (slontis) [inline apps/cmp.c:1870]: does this need more logic to handle portnum?

File: doc/man3/OSSL_HTTP_transfer.pod (hunk 1 above)
Comment (DDvO) [reply doc/man3/OSSL_HTTP_transfer.pod]: Changed

File: doc/man3/OSSL_HTTP_transfer.pod (hunk 2 above)
Comment (DDvO) [reply doc/man3/OSSL_HTTP_transfer.pod]: I've re-phrased this to:
```
The host component may be a DNS name or an IP address
where IPv6 addresses should be enclosed in square brackets C<[> and C<]>.
```

File: doc/man3/OSSL_HTTP_transfer.pod (hunk 3 above)
Comment (DDvO) [reply doc/man3/OSSL_HTTP_transfer.pod]: Changed this, see below.

File: doc/man3/OSSL_HTTP_transfer.pod (hunk 4 above)
Comment (DDvO) [reply doc/man3/OSSL_HTTP_transfer.pod]: Changed this as well, see below.

File: doc/man3/OSSL_HTTP_transfer.pod (hunk 5 above)
Comment (DDvO) [reply doc/man3/OSSL_HTTP_transfer.pod]: I've re-arranged the sentences to:
```
A path string returned is guaranteed to begin with a C</> character.
On success the function assigns via each non-NULL result pointer argument
the respective url component. On error, string results are assigned to NULL,
else they are guaranteed to be non-NULL pointers to copies of the strings.
It is the reponsibility of the caller to free them using L<OPENSSL_free(3)>.
```
which should read better now :)

File: crypto/http/http_lib.c (hunk 6 above)
Comment (DDvO) [reply crypto/http/http_lib.c]: I've added a comment explaining it:
```
/* https_len is used below as an upper bound of the prefix length */
```

File: crypto/http/http_lib.c (hunk 7 above)
Comment (DDvO) [reply crypto/http/http_lib.c:29]: This is because I use below `strtol()`, which is preferred over `a2int` etc. for its better error signaling.
A specific port range check is then needed anyway, an on success a cast to `int` can be done safely.

File: crypto/http/http_lib.c (hunk 8 above)
Comment (DDvO) [reply crypto/http/http_lib.c:55]: For HTTP(S) URLs only "http://" or "https://" may be used.
This PR does not change this behavior but makes it explicit in the documentation.

File: crypto/http/http_lib.c (hunk 9 above)
Comment (DDvO) [reply crypto/http/http_lib.c]: Done (using just more compact texts fitting in the line limit).

File: crypto/http/http_lib.c (hunk 10 above)
Comment (DDvO) [reply crypto/http/http_lib.c:115]: I've added a comment explaining this:
```
    /* cannot use p + 1 because *p is '\0' and path must start with '/' */
```

File: apps/cmp.c (hunk 11 above)
Comment (DDvO) [reply apps/cmp.c:1883]: No more logic needed here because the app uses the `&portnum` argument when calling
```
OSSL_HTTP_parse_url(opt_server, &server, &port, &portnum, &path, &ssl
```
which guarantees on success that the port string and number is valid.

**Comment by DDvO** [comment]
Thanks a lot @slontis for your good comments on this PR.
I've meanwhile handled/answered all of them.

File: crypto/http/http_lib.c (hunk 8 above)
Comment (slontis) [reply crypto/http/http_lib.c:55]: There is nothing **in the existing code** to stop you putting in a bogus URL as I have done - and it will just skip it - so the question is should it do this a bit better?

File: crypto/http/http_lib.c (hunk 10 above)
Comment (slontis) [reply crypto/http/http_lib.c:115]: ??? If you dont set it to '\0' above the if then you dont have this problem..
i.e without you setting it it will be either '\' or '\0'
if it is '\' then you can then do as I have above otherwise it is at the end.

File: crypto/http/http_lib.c (hunk 6 above)
Comment (slontis) [reply crypto/http/http_lib.c]: My point was that it is comparing the length of 2 fixed strings.. Unless someone plans on redefining these strings this seems like a strange thing to check.

File: doc/man3/OSSL_HTTP_transfer.pod (hunk 5 above)
Comment (slontis) [reply doc/man3/OSSL_HTTP_transfer.pod]: What about something like this - that still hurts my brain..

If the parameters I<phost>, I<pport> and I<ppath> are non-NULL then the returned strings must be freed by the caller using L<OPENSSL_free(3)>. If an error occurs then the returned strings will be NULL.
If a path string I<pppath> is returned it is guaranteed to begin with a C</> character.

**Review by slontis — COMMENTED** [review]
It would be good if OSSL_HTTP_parse_url() had some specific tests..

File: crypto/http/http_lib.c (hunk 8 above)
Comment (DDvO) [reply crypto/http/http_lib.c]: Good point.
I've meanwhile made sure that only the schemes "http" and "https" are allowed.
Doing so I was able to actually simplify the code.
Moreover, I've added various tests for `OSSL_HTTP_parse_url()`.

File: crypto/http/http_lib.c (hunk 10 above)
Comment (DDvO) [reply crypto/http/http_lib.c:107]: Yes, but what you are missing is that even if `*p = '\0'` there is still the problem that
the path output needs needs to be guaranteed to start with the preceding `/`
but in your code it starts with the character thereafter, and
we cannot both replace the `/` by `\` and keep the  `/` in the same piece of memory.

**Comment by DDvO** [comment]
> It would be good if OSSL_HTTP_parse_url() had some specific tests..

I felt this need, too. Done.

File: crypto/http/http_lib.c (hunk 6 above)
Comment (DDvO) [reply crypto/http/http_lib.c]: Yeah, but meanwhile I found that the whole `https_len` business is not needed any more 
after I strengthened the scheme parsing.

File: doc/man3/OSSL_HTTP_transfer.pod (hunk 5 above)
Comment (DDvO) [reply doc/man3/OSSL_HTTP_transfer.pod]: Your text suggestion does not include the statement that on success the strings returned as guaranteed to be non-NULL.
Yet I see your point to better explicitly give the parameter names - done this, and the section now reads as:
```
On success the function assigns via each non-NULL result pointer argument
I<phost>, I<pport>, I<pport_num>, I<ppath>, and I<pssl>
the respective url component.
On error, B<*phost>, B<*pport>, and B<*ppath> are assigned to NULL,
else they are guaranteed to contain non-NULL string pointers.
It is the reponsibility of the caller to free them using L<OPENSSL_free(3)>.
A string returned via B<*ppath> is guaranteed to begin with a C</> character.
```
On this occasion I replaced in the file all parameter references `B<...>` by `I<...>`.

File: crypto/http/http_lib.c [hunk 12]
```diff
@@ -50,26 +46,22 @@ int OSSL_HTTP_parse_url(const char *url, char **phost, char **pport,
     if ((buf = OPENSSL_strdup(url)) == NULL)
         goto err;
 
-    /* check for optional http[s]:// prefix */
-    p = strchr(buf, ':');
-    if (p == NULL || (size_t)(p - buf) > https_len) {
+    /* check for optional prefix "http[s]://" */
+    p = strstr(buf, "://");
```
Comment (slontis) [inline crypto/http/http_lib.c:50]: much better!

File: doc/man3/OSSL_HTTP_transfer.pod [hunk 13]
```diff
@@ -180,37 +180,39 @@ After disconnect the modified BIO will be deallocated using BIO_free_all().
...
-A path string returned is guaranteed to begin with a C</> character.
 On success the function assigns via each non-NULL result pointer argument
-the respective url component. On error, string results are assigned to NULL,
-else they are guaranteed to be non-NULL pointers to copies of the strings.
+I<phost>, I<pport>, I<pport_num>, I<ppath>, and I<pssl>
+the respective url component.
+On error, B<*phost>, B<*pport>, and B<*ppath> are assigned to NULL,
+else they are guaranteed to contain non-NULL string pointers.
 It is the reponsibility of the caller to free them using L<OPENSSL_free(3)>.
+A string returned via B<*ppath> is guaranteed to begin with a C</> character.
```
Comment (slontis) [inline doc/man3/OSSL_HTTP_transfer.pod:209]: Not sure why you are using B on the <*variable>

File: test/http_test.c [hunk 14]
```diff
@@ -151,6 +151,72 @@ static int test_http_x509(int do_get)
     return res;
 }
 
+static int test_http_url_ok(const char *url, const char *exp_host, int exp_ssl)
+{
```
Comment (slontis) [inline test/http_test.c:155]: Thanks for adding some tests

File: doc/man3/OSSL_HTTP_transfer.pod (hunk 13 above)
Comment (DDvO) [reply doc/man3/OSSL_HTTP_transfer.pod:209]: I did this here (and elsewhere) to emphasize that this is not a parameter name but an expression involving a parameter name.

**Review by slontis — APPROVED** [review]
Apart from a NIT query about the use of B<*> this looks fine.

File: test/http_test.c (hunk 14 above)
Comment (DDvO) [reply test/http_test.c:155]: Yeah, those were really missing, and having added them gives a good direct feedback on the correctness
using `make test TESTS=test_http`
(while `make test TESTS="test_cmp_cli" OPENSSL_CMP_ASPECTS=connection` 
already contained some tests for the URL parsing).

**Comment by DDvO** [comment]
> Apart from a NIT query about the use of B<*> this looks fine.

Very pleased that you approved this that swiftly
and that we have been able to further improve `OSSL_HTTP_parse_url()`.
 
Regarding the use of `B<*ppath>` I am not aware of clear "code documentation markup rules" covering such cases where not just a parameter name but some more complex expression is given.

**Comment by DDvO** [comment]
Merged - thanks!

**Comment by mtrojnar** [comment]
Changing the number of parameters for OCSP_parse_url() breaks the public API.  Was it intended?

**Comment by DDvO** [comment]
> Changing the number of parameters for OCSP_parse_url() breaks the public API. Was it intended?

This was not intended but happened indirectly, changing the related function `OSSL_HTTP_parse_url()`.
I've just submitted a fix for restoring backward compatibility: https://github.com/openssl/openssl/pull/13077.
